decision_proceduret API with assumptions

kroening opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
new decision_proceduret API

This changes the API offered by decision_proceduret to solving under
assumptions.  Solving under assumptions has been popularised by MiniSat, and
is a state-less alternative to context stacks.

This change mimics the transition from check-sat to check-sat-assuming that
SMT-LIB is undergoing.

martin-cs reviewed (approved):
Seems like the sensible thing to do.

kroening committed:
fix nondeterministic hex_trace test

The cbmc/hex_trace test pattern relies on a particular value (0) for an
uninitialized local variable.  This removes the pattern.

See
https://github.com/diffblue/cbmc/actions/runs/6672135313/job/18269948126?pr=7979
as an exemplar where this was triggered.

tautschnig commented on src/solvers/decision_procedure.cpp:
```diff
@@ -11,13 +11,21 @@ Author: Daniel Kroening, kroening@kroening.com
 {
-  return dec_solve();
+  return dec_solve(nil_exprt());
+}
```
My preference would be to use std::optional at the interface rather than nil.

kroening replied:
I was reluctant to do so since that can't avoid copying the expression.

tautschnig replied:
Wouldn't `const std::optional<exprt> &` do the trick?

kroening replied:
No, you'd still need to copy the expression into the optional when creating the optional.

kroening replied:
`std::optional< std::reference_wrapper<exprt> >` would work, but then it's getting more verbose.
